src/vmm/src/controller.rs

```diff
@@ -225,4 +227,243 @@ impl VmmController {
 
         Ok(())
     }
+
+    /// Retrieves the KVM dirty bitmap for each of the guest's memory regions.
+    pub fn get_dirty_bitmap(
+        &self,
+    ) -> std::result::Result<HashMap<usize, Vec<u64>>, VmmActionError> {
+        let mut bitmap: HashMap<usize, Vec<u64>> = HashMap::new();
+        let vmm_handle = self.vmm.lock().unwrap();
+        vmm_handle.guest_memory.with_regions_mut(
+            |slot: usize, region: &GuestRegionMmap| -> std::result::Result<(), VmmActionError> {
+                let bitmap_region = vmm_handle
+                    .vm
+                    .fd()
+                    .get_dirty_log(slot as u32, region.len() as usize)
+                    .map_err(|e| VmmActionError::InternalVmm(super::Error::DirtyBitmap(e)))?;
+                bitmap.insert(slot, bitmap_region);
+                Ok(())
+            },
+        )?;
```

Comment:
I would add this functionality as a `struct Vmm` method. The `VmmController` is just a wrapper over `Vmm` to translate between vmm interface and upper layers.

Getting dirty pages sounds like functionality worthy of being part of the `Vmm` programmatic interface.

Reply:
I moved the implementation and the unit tests out of the `controller` to the `vmm`. I left the `controller` action and entry point, but I'm considering whether we'll need them. Without an HTTP API, I think not.